clang/test/dpct/helper_files_ref/include/dpl_extras/algorithm.h

```diff
@@ -1208,6 +1349,48 @@ inline void sort_keys(
   }
 }
 
+template <typename _ExecutionPolicy, typename key_t, typename value_t,
+          typename OffsetIteratorT>
+inline void segmented_sort_pairs(
+    _ExecutionPolicy &&policy, key_t keys_in, key_t keys_out, value_t values_in,
+    value_t values_out, int64_t n, int64_t nsegments,
+    OffsetIteratorT begin_offsets, OffsetIteratorT end_offsets,
+    bool descending = false, int begin_bit = 0,
+    int end_bit = sizeof(typename ::std::iterator_traits<key_t>::value_type) *
+                  8) {
+  int compute_units =
+      policy.queue()
+          .get_device()
+          .template get_info<sycl::info::device::max_compute_units>();
+  auto sg_sizes = policy.queue()
+                      .get_device()
+                      .template get_info<sycl::info::device::sub_group_sizes>();
+  int subgroup_size = sg_sizes.empty() ? 1 : sg_sizes.back();
+  int compute_unit_threads =
+      policy.queue().get_device().is_gpu() ? subgroup_size : 1;
+  // parallel for of serial sorts when we have sufficient number of segments for
+  // load balance
+  if (nsegments > compute_units * compute_unit_threads) {
+    dpct::internal::segmented_sort_pairs_by_parallel_for_of_sorts(
+        ::std::forward<_ExecutionPolicy>(policy), keys_in, keys_out, values_in,
+        values_out, n, nsegments, begin_offsets, end_offsets, descending,
+        begin_bit, end_bit);
+  } else if (nsegments < 500) // for loop of parallel sorts when we have a small
+                              // number of total sorts to limit total overhead
+  {
+    dpct::internal::segmented_sort_pairs_by_parallel_sorts(
+        ::std::forward<_ExecutionPolicy>(policy), keys_in, keys_out, values_in,
+        values_out, n, nsegments, begin_offsets, end_offsets, descending,
+        begin_bit, end_bit);
+  } else // decent catch using 2 full sorts
```

Comment:
is is except that if compute_units * compute_unit_threads<=500, the last else{} is dead code?

Reply:
The code is written to be general purpose, and attempt to pick a reasonable algorithm based on the device targeted, on some target devices, it is true that this code may not be reachable.